core/lib/contract_verifier/src/etherscan/mod.rs

```diff
@@ -0,0 +1,391 @@
+                                pause_duration,
+                                api_error
+                            );
+                            tokio::time::sleep(pause_duration).await;
+                            self.reset_etherscan_verification_processing_started_at(
+                                verification_request.id,
+                            )
+                            .await?;
+                            started_at = Instant::now();
+                        }
+                        ApiErrorRetryPolicy::IndefinitelyRetryable { pause_duration } => {
+                            tracing::warn!(
+                                "Pausing processing for {:?} due to the error: {}",
+                                pause_duration,
+                                api_error
+                            );
+                            tokio::time::sleep(pause_duration).await;
+                            self.reset_etherscan_verification_processing_started_at(
+                                verification_request.id,
+                            )
+                            .await?;
+                            started_at = Instant::now();
+                        }
+                    }
+                }
+                Err(VerifierError::ProcessingError(processing_error)) => {
+                    break Err(processing_error.into())
+                }
+            }
+        };
+
+        match processing_result {
+            Ok(_) => {
+                self.save_result(verification_request.id).await?;
+            }
+            Err(err) => {
+                self.save_failure(verification_request.id, err.to_string())
+                    .await?;
+            }
+        }
+
+        API_CONTRACT_VERIFIER_METRICS
+            .etherscan_request_processing_time
+            .observe(started_at.elapsed());
+
+        Ok(())
+    }
+
+    pub async fn run(mut self, mut stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        loop {
+            if *stop_receiver.borrow() {
+                tracing::warn!("Stop signal received, shutting down etherscan verifier");
+                return Ok(());
+            }
+
+            // Update solc long versions, required for Etherscan verification.
+            // Internally cached, so it's fine to call it on every iteration.
+            if let Err(err) = self.solc_versions_fetcher.update_versions().await {
+                tracing::error!("Failed to update solc versions: {}", err);
+            }
+
+            // Errors are handled internally so it's safe to ignore them here
+            // to keep the processing loop running.
+            self.process_next_request().await.ok();
+
+            let wait_interval = Self::wait_for_request_interval();
+            tracing::trace!("Pausing execution for {} sec.", wait_interval.as_secs());
+            tokio::time::timeout(wait_interval, stop_receiver.changed())
+                .await
+                .ok();
```

Comment:
Note that you react to the stop signal here, but not in other places where `tokio::sleep` is called. One way to get around it is to add `TaskCanceled` to `VerifierError` and change `self.process_next_request().await.ok()` to something like

```rust
if let Err(VerifierError::Canceled) = self.process_next_request(&mut stop_receiver).await {
  continue;
}
```

Reply:
Fixed